src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+                    "Error converting i64 to i32 (for date): {:?}",
+                    e
+                ))
+            }),
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::TIMESTAMP => decode_udf_result_primitive_array::<
+            arrow::datatypes::TimestampNanosecondType,
+        >(encoded_results, &|v| {
+            v.as_i64().ok_or(DataFusionError::Internal(format!(
+                "Expected to find i64 value, but received {:?} instead",
+                v
+            )))
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::BOOLEAN => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                decoded_results.push(Some(i.as_bool().ok_or(
+                    DataFusionError::Internal(format!(
+                        "Expected to find bool value, received {:?} instead",
+                        &i
+                    )),
+                )?));
+            }
+            Ok(Arc::new(
+                decoded_results
+                    .iter()
+                    .collect::<arrow::array::BooleanArray>(),
+            ) as ArrayRef)
+        }
+
+        CreateFunctionDataType::F64 | CreateFunctionDataType::DOUBLE => {
+            decode_udf_result_primitive_array::<arrow::datatypes::Float64Type>(
+                encoded_results,
+                &|v| {
+                    v.as_f64().ok_or(DataFusionError::Internal(format!(
+                        "Expected to find f64 value, but received {:?} instead",
+                        v
+                    )))
+                },
+            )
+            .map(|a| Arc::new(a) as ArrayRef)
+        }
+        CreateFunctionDataType::F32
+        | CreateFunctionDataType::REAL
+        | CreateFunctionDataType::FLOAT => decode_udf_result_primitive_array::<
+            arrow::datatypes::Float32Type,
+        >(encoded_results, &|v| match v {
+            Value::F32(n) => Ok(*n),
+            _ => Err(DataFusionError::Internal(format!(
+                "Expected to find f64 value, but received {:?} instead",
+                v
+            ))),
+        })
+        .map(|a| Arc::new(a) as ArrayRef),
+        CreateFunctionDataType::DECIMAL {
+            precision: _,
+            scale: _,
+        } => {
+            let mut decoded_results = Vec::with_capacity(encoded_results.len());
+            for i in encoded_results {
+                let s = i.as_str().ok_or(DataFusionError::Internal(format!(
+                    "Expected to find str value, received {:?} instead",
+                    &i
+                )))?;
+                decoded_results.push(Some(s.parse::<i128>().map_err(|e| {
+                    DataFusionError::Internal(format!(
+                        "Error parsing string to i128: {:?}",
+                        e
+                    ))
+                })?));
+            }
```

Comment:
This is definitely untested :joy: I'm happy with your variant if it works. Though I'm not sure about how this decimal type even works. Do the `precision`/`scale` values matter here and should the guest function know them in order to be able to decode the i128 into an actual "decimal" value it can do work on?

```suggestion
            let decoded_results = encoded_results.iter().map(|i|
                i.as_str().ok_or(DataFusionError::Internal(format!(
                    "Expected to find str value, received {:?} instead",
                    &i
                ))).and_then(|s| s.parse::<i128>().map_err(|e|
                    DataFusionError::Internal(format!(
                        "Error parsing string to i128: {:?}",
                        e
                )))))
            .collect::<Result<arrow::array::Decimal128Array>>()?;
            Ok(Arc::new(decoded_results) as ArrayRef)


```

Reply:
I'm not entirely sure. It's telling that's `Decimal128Type::Native = i128` and there's no clear logic involving precision and scale in that file. My guess is that it's only display related (but I could be wrong ofc).

Reply:
Looks like the precision/scale do have semantic meaning:

https://arrow.apache.org/docs/python/generated/pyarrow.decimal128.html

The scale is basically by how many digits the decimal is shifted:

> As an example, decimal128(7, 3) can exactly represent the numbers 1234.567 and -1234.567 (encoded internally as the 128-bit integers 1234567 and -1234567, respectively), but neither 12345.67 nor 123.4567.
> decimal128(5, -3) can exactly represent the number 12345000 (encoded internally as the 128-bit integer 12345), but neither 123450000 nor 1234500.